demos/adjoint/adjoint_demo.py

```diff
@@ -0,0 +1,348 @@
+# import numpy as np
+# Delta_temp = Function(Tic.function_space(), name="Delta_Temperature")
+# Delta_temp.dat.data[:] = np.random.random(Delta_temp.dat.data.shape)
+#
+# # Perform the Taylor test to verify the gradients
+# minconv = taylor_test(reduced_functional, Tic, Delta_temp)
+# -
+
+# The `taylor_test` function computes the Taylor remainder and verifies that the convergence rate is close to the theoretical value of $O(2.0)$. This ensures that our gradients are accurate and reliable for optimisation.
+
+# Running the inversion
+# ------------------------
+# In the final section, we run the optimisation method. First, we define lower and upper bounds for the optimisation problem to guide the optimisation method towards a more constrained solution.
+#
+# For this simple problem, we perform a bounded nonlinear optimisation where the temperature is only permitted to lie in the range [0, 1]. This means that the optimisation problem should not search for solutions beyond these values.
+
+# +
+# Define lower and upper bounds for the temperature
+T_lb = Function(Tic.function_space(), name="Lower Bound Temperature")
+T_ub = Function(Tic.function_space(), name="Upper Bound Temperature")
+
+# Assign the bounds
+T_lb.assign(0.0)
+T_ub.assign(1.0)
+
+# Define the minimisation problem, with the goal to minimise the reduced functional
+# Note: in some scenarios, the goal might be to maximise (rather than minimise) the functional.
+minimisation_problem = MinimizationProblem(reduced_functional, bounds=(T_lb, T_ub))
+# -
+
+# Using the LinMore Optimiser
+# ---------------------------
+#
+# In this tutorial, we employ the trust region method of Lin and Moré (1999) implemented in ROL (Rapid Optimization Library).
+# Lin-Moré is a truncated Newton method, which involves the repeated application of an iterative algorithm to approximately solve Newton’s equations (Dembo and Steihaug, 1983).
+#
+# Lin-Moré effectively handles provided bound constraints by ensuring that variables remain within their specified bounds.
+# During each iteration, variables are classified into "active" and "inactive" sets. Variables at their bounds that do not
+# allow descent are considered active and are fixed during the iteration. The remaining variables, which can change without
+# violating the bounds, are inactive. These properties make the algorithm a robust and efficient method for solving bound-constrained
+# optimisation problems.
+#
+# A notable feature of this optimisation approach in ROL is its checkpointing capability. For every iteration, all information necessary to restart the optimisation from that iteration is saved in the specified `checkpoint_dir`.
+
+# +
+# Define the LinMore Optimiser class with checkpointing capability
+optimiser = LinMoreOptimiser(
+    minimisation_problem,
+    minimisation_parameters,
+    checkpoint_dir="optimisation_checkpoint",
+)
+
+# Restore from last possible checkpoint
+# optimiser.restore()
+# Run the optimisation
+optimiser.run()
```

Comment:
It feels like an abrupt end to a REALLY great tutorial. Why not plot the initial guess as a function of iteration count? Why not also tell the user how to access the "misfit" terms, and plot those misfits (objective functional and initial state) as a function of iteration?

Reply:
How long does this last step take to run for you? I had to stop it running at my end, as I need to get home. Again, perhaps worth telling the user roughly how long it is expected to take. And definitley need something to close the tutorial, as I suggested above. Otherwise - they have an optimised solution but they don't know what it looks like and how to plot various things associated with that optimisation procedure.